src/compiler/compiler.rs

```diff
@@ -275,124 +296,205 @@ impl Compiler {
-                        debug!("[{}]: Compiled but not cacheable", out_file);
-                        Ok((CompileResult::NotCacheable, compiler_result))
-                    }
+                me.compile(&creator,
+                           preprocessor_result,
+                           parsed_args,
+                           &cwd,
+                           pool,
+                           outputs,
+                           storage,
+                           key,
+                           miss_type)
+            }))
+        }))
+    }
+
+    fn compile<T>(&self,
+                  creator: &T,
+                  preprocessor_result: process::Output,
+                  parsed_args: ParsedArguments,
+                  cwd: &str,
+                  pool: CpuPool,
+                  outputs: HashMap<String, PathBuf>,
+                  storage: Arc<Storage>,
+                  key: String,
+                  miss_type: MissType)
+                  -> Box<Future<Item=(CompileResult, process::Output),
+                                Error=io::Error>>
+        where T: CommandCreatorSync,
+    {
+        let process::Output { stdout, .. } = preprocessor_result;
+        let start = Instant::now();
+
+        let compile = self.kind.compile(creator, self, stdout, &parsed_args, cwd, &pool);
+        Box::new(compile.and_then(move |(cacheable, compiler_result)| {
+            let duration = start.elapsed();
+            if compiler_result.status.success() {
+                if cacheable == Cacheable::Yes {
+                    debug!("[{}]: Compiled, storing in cache",
+                           parsed_args.output_file());
+                    // fall through
                 } else {
-                    debug!("[{}]: Compiled but failed, not storing in cache", out_file);
-                    Ok((CompileResult::CompileFailed, compiler_result))
+                    // Not cacheable
+                    debug!("[{}]: Compiled but not cacheable",
+                           parsed_args.output_file());
+                    return future::ok((CompileResult::NotCacheable, compiler_result)).boxed()
+                        as Box<Future<Item=_, Error=_>>
                 }
+            } else {
+                debug!("[{}]: Compiled but failed, not storing in cache",
+                       parsed_args.output_file());
+                return future::ok((CompileResult::CompileFailed, compiler_result)).boxed()
             }
-        }
+            let mut entry = match storage.start_put(&key) {
+                Ok(entry) => entry,
+                Err(e) => return future::err(e).boxed()
+            };
+            let write = pool.spawn_fn(move || -> io::Result<_> {
+                for (key, path) in &outputs {
+                    let mut f = try!(File::open(&path));
+                    try!(entry.put_object(key, &mut f).map_err(|e| {
+                        let msg = format!("failed to put object `{:?}` in \
```

Comment:
You lost the `out_file` from the error message here, which is not a huge problem but I put those in all the log messages because otherwise trying to interpret the sccache log from a real parallel build was a nightmare.